### PR TITLE
[SPARK-44641][SQL] Incorrect result in certain scenarios when SPJ is not triggered 

### PR DESCRIPTION
**Pull request by sunchao**
### What changes were proposed in this pull request?

This PR makes sure we use unique partition values when calculating the final partitions in `BatchScanExec`, to make sure no duplicated partitions are generated.

### Why are the changes needed?

When `spark.sql.sources.v2.bucketing.pushPartValues.enabled` and `spark.sql.sources.v2.bucketing.partiallyClusteredDistribution.enabled` are enabled, and SPJ is not triggered, currently Spark will generate incorrect/duplicated results. 

This is because with both configs enabled, Spark will delay the partition grouping until the time it calculates the final partitions used by the input RDD. To calculate the partitions, it uses partition values from the `KeyGroupedPartitioning` to find out the right ordering for the partitions. However, since grouping is not done when the partition values is computed, there could be duplicated partition values. This means the result could contain duplicated partitions too.

### Does this PR introduce _any_ user-facing change?

No, this is a bug fix.

### How was this patch tested?

Added a new test case for this scenario.

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala [hunk 1]
```diff
@@ -1039,4 +1039,60 @@ class KeyGroupedPartitioningSuite extends DistributionAndOrderingSuiteBase {
       }
     }
   }
+
+  test("SPARK-44641: duplicated records when SPJ is not triggered") {
```
Comment (dongjoon-hyun) [inline sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala:1043]: May I ask which Spark versions are affected, @sunchao ? (Also, cc @aokolnychyi )

Comment (sunchao) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala:1043]: @dongjoon-hyun this affects Spark 3.4.0 and 3.4.1

Comment (sunchao) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala:1043]: it'd be great if we can get this in Spark 3.5.0

**Comment by sunchao** [comment]
cc @viirya @szehon-ho (who discovered this bug) @cloud-fan too

File: sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala (hunk 1 above)
Comment (dongjoon-hyun) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala:1043]: +1 for backporting to branch-3.5.

BTW, could you make it as `Blocker` to achieve that?

Comment (sunchao) [reply sql/core/src/test/scala/org/apache/spark/sql/connector/KeyGroupedPartitioningSuite.scala:1043]: Sure, it's done

**Review by szehon-ho — COMMENTED** [review]
Fix looks good to me.  

For my knowledge, curious why we need to run all this code block in BatchScanExec::inputRDD (starting line 140) if SPJ not triggered?

**Review by szehon-ho — APPROVED** [review]

**Comment by sunchao** [comment]
Thanks @szehon-ho !

> For my knowledge, curious why we need to run all this code block in BatchScanExec::inputRDD (starting line 140) if SPJ not triggered?

ATM even though SPJ is not triggered, the scan partitioning is still `KeyGroupedPartitioning` and therefore need to be handled accordingly.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 2]
```diff
@@ -355,6 +355,13 @@ case class KeyGroupedPartitioning(
 
   override def createShuffleSpec(distribution: ClusteredDistribution): ShuffleSpec =
     KeyGroupedShuffleSpec(this, distribution)
+
+  def uniquePartitionValues: Seq[InternalRow] = {
```
Comment (szehon-ho) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:359]: Actually one thought, would it make sense to make it lazy?  It seems only used in this one specific case when both flags are on.

Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:359]: Yes, we can probably make it lazy val

**Comment by viirya** [comment]
I will review this today or in weekend.

**Review by cloud-fan — APPROVED** [review]
good catch!

**Review by viirya — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala [hunk 3]
```diff
@@ -355,6 +355,13 @@ case class KeyGroupedPartitioning(
...
+    partitionValues
+        .map(InternalRowComparableWrapper(_, expressions))
+        .distinct
+        .map(_.row)
+  }
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:364]: Hmm, one question. Based on `KeyGroupedPartitioning` document, `partitionValues` must be unique? This issue is caused by duplicated partition values? Is the document wrong?

Comment (sunchao) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/physical/partitioning.scala:364]: Good point. This document is stale. I just updated it.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM.

**Comment by sunchao** [comment]
Thanks! merged to master/branch-3.4/branch-3.5

**Pull request closed without merging**